### PR TITLE
feat (metadata): add basic model setup for customer metadata

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Creating a customer with Lago’s basic information is not enough to bring context.

## Description

This PR adds customer metadata table and model relations/validations

### CODE REVIEW

File: spec/models/metadata/customer_metadata_spec.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+require 'rails_helper'
+
+RSpec.describe Metadata::CustomerMetadata, type: :model do
+  subject(:metadata) { described_class.new(attributes) }
+
+  let(:customer) { create(:customer) }
+  let(:key) { 'hello' }
+  let(:value) { 'abcdef' }
+  let(:attributes) do
+    { key: key, value: value, customer: customer, display_in_invoice: true }
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:12]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/models/metadata/customer_metadata_spec.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+  describe 'validations' do
+    context 'when key is unique' do
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:16]: RSpec/RepeatedExampleGroupBody: Repeated context block body on line(s) [32, 46] (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/RepeatedExampleGroupBody)

File: spec/models/metadata/customer_metadata_spec.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      it 'validates the key' do
+        expect(metadata).to be_valid
+      end
+    end
+
+    context 'when key is not unique' do
+      let(:old_metadata) { create(:customer_metadata, customer: customer, key: 'hello') }
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:23]: Style/HashSyntax: Omit the hash value. (https://rubystyle.guide#hash-literals)

File: spec/models/metadata/customer_metadata_spec.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+      before { old_metadata }
+
+      it 'validates the key' do
+        expect(metadata).not_to be_valid
+      end
+    end
+
+    context 'when key length is valid' do
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:32]: RSpec/RepeatedExampleGroupBody: Repeated context block body on line(s) [16, 46] (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/RepeatedExampleGroupBody)

File: spec/models/metadata/customer_metadata_spec.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    context 'when key length is not valid' do
+      let(:key) { 'hello-hello-hello-hello-hello' }
+
+      it 'validates the key' do
+        expect(metadata).not_to be_valid
+      end
+    end
+
+    context 'when value length is valid' do
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:46]: RSpec/RepeatedExampleGroupBody: Repeated context block body on line(s) [16, 32] (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/RepeatedExampleGroupBody)

File: db/migrate/20230221070501_create_customer_metadata.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class CreateCustomerMetadata < ActiveRecord::Migration[7.0]
+  def change
+    create_table :customer_metadata, id: :uuid do |t|
+      t.references :customer, type: :uuid, null: false, foreign_key: true, index: true
+
+      t.string :key, null: false
```
Comment (vincent-pochet) [inline db/migrate/20230221070501_create_customer_metadata.rb:8]: What about a unique index on both `customer_id` and `key` fields to match the uniqueness validation?

**Review by vincent-pochet — COMMENTED** [review]
Look great!
Only one comment for uniqueness enforcement at DB level

File: spec/models/metadata/customer_metadata_spec.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      let(:old_metadata) { create(:customer_metadata, customer:, key: 'hello') }
+
+      before { old_metadata }
+
+      it 'validates the key' do
+        expect(metadata).not_to be_valid
+      end
+    end
+
+    context 'when length constraint is satisfied' do
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:32]: RSpec/RepeatedExampleGroupBody: Repeated context block body on line(s) [16] (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/RepeatedExampleGroupBody)

File: spec/models/metadata/customer_metadata_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    context 'when key is not unique' do
+      let(:old_metadata) { create(:customer_metadata, customer:, key: 'hello') }
+
+      before { old_metadata }
+
+      it { expect(metadata).not_to be_valid }
+    end
+
+    context 'when length constraint is satisfied', :tag do
```
Comment (jdenquin) [inline spec/models/metadata/customer_metadata_spec.rb:28]: RSpec/RepeatedExampleGroupBody: Repeated context block body on line(s) [16] (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/RepeatedExampleGroupBody)

File: db/migrate/20230221070501_create_customer_metadata.rb (hunk 6 above)
Comment (lovrocolic) [reply db/migrate/20230221070501_create_customer_metadata.rb:8]: Make sense to add it on DB level as well! Added!

**Review by vincent-pochet — APPROVED** [review]

**Pull request merged**